### ISSUE TITLE
Gradient on clipping path lost when optimizing with scour

### ISSUE THREAD

**Issue by felixbuenemann** [scour-project/scour#66]
The following file looses the gradient used to fill the clipping path on the outer circle when optimized with scour.

Preview: https://www.dropbox.com/s/p3axbgr9ek59gcu/gradientproblem.svg?dl=0
Download: https://www.dropbox.com/s/p3axbgr9ek59gcu/gradientproblem.svg?dl=1

Inspecting the scour svg shows that the gradient itself (liner0) is still present and used to fill the clip path, but the circle is rendered a solid white.

I tried optimizing the SVG with `svgo` instead and although it did similar optimizations it didn't break the gradient.

Screenshot (original left, scour-optimized right):

![](https://www.dropbox.com/s/r53kbfl64ykl1in/Screenshot%202016-08-22%2022.34.29.png?dl=1)

The SVG was generated from a PDF file using poppler and cairo's svg renderer.

**Comment by Ede123** [comment]
I've tracked down the issue, which is some confusion regarding the default value of the [`x2` attribute](https://www.w3.org/TR/SVG/pservers.html#LinearGradientElementX2Attribute). *(If the attribute is not specified, the effect is as if a value of '100%' were specified.)*

Consider the following minimal example:
``` SVG
<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="128" height="128" viewBox="0 0 128 128">
 <defs>
  <linearGradient id="linear0" gradientTransform="scale(128)" gradientUnits="userSpaceOnUse" x2="1">
   <stop offset="0" stop-color="blue"/>
   <stop offset="1" stop-color="red"/>
  </linearGradient>
 </defs>
 <rect width="128" height="128" fill="url(#linear0)"/>
</svg>
```

* Scour removes the part `x2="1"` from the `linearGradient` as it assumes 1=100% and therefore the default value.
* In this case I *think* it should be 1=1px, though, in which case the gradient would be scaled (correctly) to a width of 128px.
* If the value is removed 100%=128px is probably the correct value, which would get scaled (incorrectly) to a much too large 16384px which would be the reason why the gradient appears solid blue (or white in the initial example).


Interestingly *a lot* of SVG renderers treat no specification of `x2`, i.e. the default value of 100%, as if a value of 1px was specified. As a result the SVG is rendered as intended in e.g. Inkscape, GIMP and libRSVG.

Can someone check if my interpretation is correct and therefore Scour / Inkscape / GIMP / libRSVG need to be fixed or if actually those programs are correct and a bug for e.g. Firefox needs to be filed?

**Comment by Ede123** [comment]
Seems I actually [filed a bug for this issue on the Inkscape bug tracker](https://bugs.launchpad.net/inkscape/+bug/1153706) over three years ago. :smile:

**Comment by felixbuenemann** [comment]
I think the spec is pretty clear. If no unit is given, it is assumed to be the unit of the current coordinate system, which should usually be pixels. 

The x2 value is of type `<coordinate>`:

> A <coordinate> is a length in the user coordinate system that is the given distance from the origin of the user coordinate system along the relevant axis (the x-axis for X coordinates, the y-axis for Y coordinates). Its syntax is the same as that for <length>.

> For length values in SVG-specific properties and their corresponding presentation attributes, the length unit identifier is optional. If not provided, the length value represents a distance in the current user coordinate system. 

So a value of 1 means 1px not 100% and cannot be removed as done here:

https://github.com/scour-project/scour/blob/57f93efc894b1ba4c30e42b549b0081185029c36/scour/scour.py#L1573

The checks for cx, cy and r seem to share the same problem.

**Comment by Ede123** [comment]
Yes, I agree... I remember now, that I looked into this over three years ago (when filing the bug for Inkscape) and also then came to the conclusion that Inkscape/GIMP/libRSVG where wrong.

**Comment by Ede123** [comment]
I just merged #70 which also fixes this issue.

Since I refactored some code and added functionality to remove some additional default attributes you're welcome to check for any regressions that might have gone unnoticed.

**Comment by felixbuenemann** [comment]
Thanks, I tried scour master on the problematic file and it worked fine.

If someone finds this issue, according to the SVG 1.1 spec the behavior depends on the value of `gradientUnits`:  If `gradientUnits` is equal to `objectBoundingBox` (the default), then the values are interpreted in the abstract coordinate system of the bounding box which ranges from `0,0` (top-left) to `1,1` (bottom-right) and if it is set to `userSpaceOnUse` they are interpreted in the current user coordinate system, usually pixels.

See: https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/gradientUnits
